File: golang/v2/ao.go
package layrzprotocol

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Based on the `Layrz Protocol v2` specification. AoPacket is the success packet
// sent from the server to the device
type AoPacket struct {
	// Is the timestamp of the response packet
	Timestamp time.Time
}

// FromPacket is a method that converts a raw packet to a PaPacket
// based on the `Layrz Protocol v2` specification
//
// Returns an error if the packet is invalid
func (p *AoPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Ao>") || !strings.HasSuffix(*raw, "</Ao>") {
		return errors.New("invalid package, should be <Ao>...</Ao>")
	}

	*raw = strings.TrimPrefix(*raw, "<Ao>")
	*raw = strings.TrimSuffix(*raw, "</Ao>")

	parts := strings.Split(*raw, ";")
	if len(parts) != 2 {
		return errors.New("invalid package, should have 2 parts")
	}

	receivedCrc, err := strconv.ParseUint(parts[1], 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := calculateCrc([]byte(fmt.Sprintf("%s;", parts[0])))

	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	rawTimestamp, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return errors.New("cannot convert timestamp to integer")
	}

	p.Timestamp = time.Unix(rawTimestamp, 0)

	return nil
}

// ToPacket is a method that converts a PaPacket to a raw packet
// based on the `Layrz Protocol v2` specification
func (p *AoPacket) ToPacket() *string {
	content := ""
	content += strconv.FormatInt(p.Timestamp.Unix(), 10) + ";"

	crc := calculateCrc([]byte(content))
	content += fmt.Sprintf("%04X", crc)

	content = fmt.Sprintf("<Ao>%s</Ao>", content)
	return &content
}

File: dart/lib/src/packets/src/server/ao.dart
part of '../../packets.dart';

class AoPacket extends ServerPacket {
  final DateTime timestamp;

  /// [AoPacket] is the ACK package.
  ///
  /// This package is part of the package sent from the server to the device.
  ///
  /// Specifically, this package is sent when the server doesn't have commands waiting for the device.
  AoPacket({
    /// [messageId] is the message ID that is being ACKed.
    /// This is identified in the package as `MSG_ID`
    required this.timestamp,
  }) : super();

  /// [fromPacket] creates a [AoPacket] from a string package in the format of `Layrz Protocol v2`.
  static AoPacket fromPacket(String raw) {
    if (!raw.startsWith('<Ao>') || !raw.endsWith('</Ao>')) {
      throw ParseException('Invalid identification package, should be <Ao>...</Ao>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    if (parts.length != 2) {
      throw MalformedException('Invalid package parts, should have 2 parts');
    }

    int? receivedCrc = int.tryParse(parts[1], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 1).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    try {
      return AoPacket(timestamp: DateTime.parse(parts[0]));
    } catch (e) {
      throw ParseException('Invalid timestamp');
    }
  }

  /// [toPacket] returns the package in the format of `Layrz Protocol v2`.
  @override
  String toPacket() {
    String payload = '${timestamp.millisecondsSinceEpoch ~/ 1000};';

    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();

    return '<Ao>$payload$crc</Ao>';
  }

  @override
  String toString() => toPacket();

  @override
  AoPacket copyWith({
    DateTime? timestamp,
  }) {
    return AoPacket(
      timestamp: timestamp ?? this.timestamp,
    );
  }
}
